shared/types/indexer.go

```diff
@@ -0,0 +1,239 @@
+		return err
+	}
+	hash, err := result.HashFromBytes(bz)
+	if err != nil {
+		return err
+	}
+	hashKey, err := d.indexByHash(hash, bz)
+	if err != nil {
+		return err
+	}
+	if err := d.indexByHeightAndIndex(result.GetHeight(), result.GetIndex(), hashKey); err != nil {
+		return err
+	}
+	if err := d.indexBySender(result.GetSender(), hashKey); err != nil {
+		return err
+	}
+	if err := d.indexByRecipient(result.GetRecipient(), hashKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *DefaultTxIndexer) GetByHash(hash []byte) (TxResult, error) {
+	return d.get(d.hashKey(hash))
+}
+
+func (d *DefaultTxIndexer) GetByHeight(height int64) ([]TxResult, error) {
+	return d.getAll(d.heightKey(height), DefaultHeightOrdering)
+}
+
+func (d *DefaultTxIndexer) GetBySender(sender string, descending bool) ([]TxResult, error) {
+	return d.getAll(d.senderKey(sender), descending)
+}
+
+func (d *DefaultTxIndexer) GetByRecipient(rec string, descending bool) ([]TxResult, error) {
+	return d.getAll(d.recipientKey(rec), descending)
+}
+
+// kv helper functions
+
+func (d *DefaultTxIndexer) getAll(prefix []byte, descending bool) (res []TxResult, err error) {
+	hashKeys, err := d.db.GetAll(prefix, descending)
+	if err != nil {
+		return nil, err
+	}
+	for _, hashKey := range hashKeys {
+		var txResult TxResult
+		txResult, err = d.get(hashKey)
+		if err != nil {
+			return
+		}
+		res = append(res, txResult)
+	}
+	return
+}
+
+func (d *DefaultTxIndexer) get(key []byte) (TxResult, error) {
+	bz, err := d.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	return new(DefaultTxResult).FromBytes(bz)
+}
+
+// index helper functions
+
+func (d *DefaultTxIndexer) indexByHash(hash, bz []byte) (hashKey []byte, err error) {
+	key := d.hashKey(hash)
+	return key, d.db.Put(key, bz)
+}
+
+func (d *DefaultTxIndexer) indexByHeightAndIndex(height int64, index int32, bz []byte) error {
+	return d.db.Put(d.heightAndIndexKey(height, index), bz)
+}
+
+func (d *DefaultTxIndexer) indexBySender(sender string, bz []byte) error {
+	return d.db.Put(d.senderKey(sender), bz)
+}
+
+func (d *DefaultTxIndexer) indexByRecipient(recipient string, bz []byte) error {
+	if recipient == "" {
+		return nil
+	}
+	return d.db.Put(d.recipientKey(recipient), bz)
+}
+
+// key helper functions
+
+func (d *DefaultTxIndexer) hashKey(hash []byte) []byte {
+	return d.key(HashPrefix, hex.EncodeToString(hash))
+}
+
+func (d *DefaultTxIndexer) heightAndIndexKey(height int64, index int32) []byte {
+	return append(d.heightKey(height), []byte(elenEncoder.EncodeInt(int(index)))...)
+}
+
+func (d *DefaultTxIndexer) heightKey(height int64) []byte {
+	return d.key(HeightPrefix, elenEncoder.EncodeInt(int(height))+"/")
+}
+
+func (d *DefaultTxIndexer) senderKey(address string) []byte {
+	return d.key(SenderPrefix, address)
+}
+
+func (d *DefaultTxIndexer) recipientKey(address string) []byte {
+	return d.key(RecipientPrefix, address)
+}
+
+func (d *DefaultTxIndexer) key(prefix rune, postfix string) []byte {
```

Comment:
For my personal learning and understanding, why are we using the `rune` type for the `prefix`?

Reply:
`rune` is a single unicode character which is like a `char` of the string